### DISCUSSION TITLE
关于XTLS流量特征是否可能导致被检测的一个问题

### DISCUSSION THREAD

**Discussion by Dawn-ATCG** [XTLS/Xray-core#593] [Q&A]
(新手小白, 如有冒犯还请海涵, 看不懂代码只能来提问)

根据描述, XTLS之所以能够更快, 是省略了对https流量的二次tls加密, 即直接使用已经加密好的https流量加密部分进行传输, 那么这个已经加密好的https流量部分, 是"明文传输"吗? 即是不是原封不动的进行传输呢?

如果是这样, 会不会导致可能存在的一种检测方式? 如果使用vps访问境内的网站, 境内网站发送的https流量, 过墙时被记录部分, 由于XTLS的特性, vps在接收后返回到国内用户过墙被记录, 还是原封不动的传输, 导致了二者流量一致, 所以可以推断这个vps进行了代理?

**Comment by RPRX** [comment]
是的，请不要使用代理访问境内 IP 网站，这是一个基本实践问题，**因为你使用任何代理访问境内网站，代理的 IP 都会被记录、上传、标记。**

这套机制已经很成熟了，根据内部人士的消息，一旦“你”使用代理访问了境内 IP，“你”就会被标记为在使用此代理（甚至还会标注情商）。

1. 因为你直连那个 VPS，并且时间吻合，所以本地 IP 被标记了。
2. 你挂着代理打开了境内应用比如微信，于是...

以上是早已被部署的检测方式，所以实践中应在服务端屏蔽所有境内 IP。

**Reply by Dawn-ATCG** [reply]
vps是我个人使用的, 所以已经尽量避免了使用代理进行访问境内网站, 不过由于本地分流策略使用的是白名单, 不可避免的有少量本是境内的网站但是通过代理访问. 那么对于这种方式, 如果使用多级代理, 我连接境外vps A 然后A连接B, B进行访问站点, 会不会最大程度减少被标记的情况?

**Reply by RPRX** [reply]
@Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。

**Reply by galaxyskyknight** [reply]
> @Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。

但这不是一个悖论吗，如果我自己的宽带没有固定的公网IP,那不是把我自己也屏蔽了？

**Reply by JeffreyZon** [reply]
> @Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。

请问大神要怎么做到这一点？

**Reply by galaxyskyknight** [reply]
大佬不要笑话我啊，是不是这样嘛.....

**Reply by ArcCal** [reply]
> 
> 
> > @Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。
> 
> 但这不是一个悖论吗，如果我自己的宽带没有固定的公网IP,那不是把我自己也屏蔽了？

怎么会屏蔽自己呢

**Reply by CelloDude** [reply]
> > @Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。
> 
> 但这不是一个悖论吗，如果我自己的宽带没有固定的公网IP,那不是把我自己也屏蔽了？

You could achieve this by using iptables:

Before you do this, you **ALWAYS need to make sure that you are not blocking yourself from connecting/SSH to the server**.

A fixed public IP address is often not possible.
So make sure you have another server that is free from iptables restrictions or only allow SSH traffic.
This way, if you accidentally blocked yourself from connecting to the iptables-restricted server, you can use this other server to connect to your iptables-restricted server instead.


You could set iptables INPUT and OUTPUT policies as ACCEPT and (maybe) flush all iptables rules first.

sudo -i

iptables -P INPUT ACCEPT
iptables -P OUTPUT ACCEPT

This next line of command is only if you don't wanna keep the old iptables rules.
**DO NOT RUN** the following line, **if you want to keep the old iptables rules**

iptables -F


1. IMPORTANT: At the top of iptables: **ACCEPT all INPUT and OUTPUT traffic from your public IP address in home country**.
You can check your public IP by visiting any "ip check" website.

iptables -I INPUT 1 -s XXX.XXX.XXX.XXX -j ACCEPT
iptables -I OUTPUT 1 -d XXX.XXX.XXX.XXX -j ACCEPT


2. DROP all INPUT and OUTPUT traffic from all home country IP addresses.
You can get a list of home country IP addresses by searching.
Or most proxy software (e.g. netch) would have a list of home country IP addresses in CIDR format.
That list is usually saved as a text file inside the software folder.

Enter every CIDR IP address in the list as iptables commands in the following format.

iptables -A INPUT -s XXX.XXX.XXX.XXX/CIDR -j DROP
iptables -A OUTPUT -d XXX.XXX.XXX.XXX/CIDR -j DROP
And repeat until the list is over.

You probably want to use a bash script to do it.
Basically, enter all lines of commands into a XXX.sh script file, and run the file
It is easier to make a script like this by using Excel.
Open the list as a comma-separated .csv file.
Add "iptables -A INPUT -s " before each CIDR IP address.
Add " -j DROP" after each CIDR IP address.
And do the same for each OUTPUT line.

Then, save the file as a comma-separated .csv file.
And change .csv to .txt
Open .txt file and replace all commas "," with nothing.
The csv file has a different type of new-line character at the end of each line.
So copy all texts in the text file, paste them into a nano session and save.

sudo -i

nano /root/XXX.sh

Use Ctrl + X to exit the nano interface
Type "y" to confirm saving, then "Enter", and done.

It is helpful to mark your progress your script by using echo.
At the beginning of script, maybe insert a line like:
echo "start"
Halfway in script, maybe insert a line like:
echo "half done"
At the end of script, maybe insert a line like:
echo "all done"

(Or if you have made the XXX.sh script file on your computer, you can upload it to /root/XXX.sh in your server.)

Type the following commands to run the script in background.

sudo -i

chmod +x XXX.sh
nohup bash XXX.sh &

And the script would run in background.
Feel free to exit the SSH session and wait for it to finish running.

After around 3 minutes (depends on your server performance), you can connect again to check.
Once you connect/SSH to your server again, you can check the output log file.

nano nohup.out

Here, you can find the echo messages from your script.
This helps you to track the progress.

Use Ctrl + X to exit the nano interface

And once the script has finished running, you have completed iptables settings.

(This section in bold would probably help answer your question.)
**NOTE: You do not need to remove your public IP address from the home country list.**
**Because iptables only cares about the first matching entry in the list.**






One question:
If using iptables to set FORWARD traffic policy as DROP, would this make the server more secure?
If someone can explain if it would or not, that would be great.
Much appreciation in advance.

**Reply by galaxyskyknight** [reply]
what does that mean? not understood. if my public IP is just in the home country list and I actively access the VPS, what is the iptables reaction if I ban the homecontry list? just I am the first match entry for a particular CIDR list, is it ignored my access? or match but permit? totally confused.
**NOTE: You do not need to remove your public IP address from the home country list.
Because iptables only cares about the first matching entry in the list.**

**Reply by galaxyskyknight** [reply]
I tried this, If I block CN zone(CIDR collection by using ipset), I cannot acess the VPS any more. so what indeed it is?

**Reply by jiuqi9997** [reply]
> 但这不是一个悖论吗，如果我自己的宽带没有固定的公网IP,那不是把我自己也屏蔽了？

你在服务端的outbound屏蔽不就好了。geoip:cn

**Reply by SimonTLS** [reply]
通过入站inbound连入服务端，在服务端出站outbound屏蔽geoip:cn

**Reply by galaxyskyknight** [reply]
好思路，已解决

**Reply by liamcos** [reply]
> 通过入站inbound连入服务端，在服务端出站outbound屏蔽geoip:cn

```
{
            "protocol": "blackhole",
            "settings": {
                "response": {
                    "type": "http"
                }
            },
            "tag": "blocked"
}

```
```
"routing": {
        "domainStrategy": "IPOnDemand",
        "rules": [
            {
                "domain": [
                    "geosite:cn"
                ],
                "outboundTag": "blocked",
                "type": "field"
            },
            {
                "ip": [
                    "geoip:cn"
                ],
                "outboundTag": "blocked",
                "type": "field"
            }
       ]
}
```
请问是这样的吗？为什么这样设置后没有作用？看记录里面已经有[blocked]字样了，但依旧可以正常访问并没有被blocked。。。
@SimonWe1
@jiuqi9997 
@galaxyskyknight

**Reply by liamcos** [reply]
不知为什么客户端设置完是有效的，在服务端就无效。比如如下设置：
```
"routing": {
        "domainStrategy": "IPOnDemand",
        "rules": [
            {
                "domain": [
                    "bilibili.com"
                ],
                "outboundTag": "blocked",
                "type": "field"
            }
       ]
}

```
客户端这样设置后就不能访问了，但在服务端同样设置后并没有被blocked依旧可以正常访问。

**Reply by galaxyskyknight** [reply]
跟你解析出的IP归属有关，我猜是这样。看你用的哪里的dns

**Reply by shenlijun** [reply]
IPOnDemand换成IPIfNonMatch的话，理论上可以屏蔽bilibili

**Reply by ghost** [reply]
> > @Dawn-ATCG 无论你的本地分流策略是什么，都应在 **服务端** 屏蔽所有境内 IP，即黑名单，这是最佳实践。
> 
> 但这不是一个悖论吗，如果我自己的宽带没有固定的公网IP,那不是把我自己也屏蔽了？

出站屏蔽策略。

**Reply by Corpama** [reply]
> > 通过入站inbound连入服务端，在服务端出站outbound屏蔽geoip:cn
> 
> ```
> {
>             "protocol": "blackhole",
>             "settings": {
>                 "response": {
>                     "type": "http"
>                 }
>             },
>             "tag": "blocked"
> }
> ```
> 
> ```
> "routing": {
>         "domainStrategy": "IPOnDemand",
>         "rules": [
>             {
>                 "domain": [
>                     "geosite:cn"
>                 ],
>                 "outboundTag": "blocked",
>                 "type": "field"
>             },
>             {
>                 "ip": [
>                     "geoip:cn"
>                 ],
>                 "outboundTag": "blocked",
>                 "type": "field"
>             }
>        ]
> }
> ```
> 
> 请问是这样的吗？为什么这样设置后没有作用？看记录里面已经有[blocked]字样了，但依旧可以正常访问并没有被blocked。。。 @SimonWe1 @jiuqi9997 @galaxyskyknight

给后面看到这个问题的人：在配置好出站的ip与site的block之后，记得更新服务端的geosite.dat与geoip.dat。亲测可以成功block。附赠一条，香港的vps，特别是阿里云国际版香港和腾讯云香港的，记得单独屏蔽轮子网站。

**Reply by nichuanfang** [reply]
> 不知为什么客户端设置完是有效的，在服务端就无效。比如如下设置：
> 
> ```
> "routing": {
>         "domainStrategy": "IPOnDemand",
>         "rules": [
>             {
>                 "domain": [
>                     "bilibili.com"
>                 ],
>                 "outboundTag": "blocked",
>                 "type": "field"
>             }
>        ]
> }
> ```
> 
> 客户端这样设置后就不能访问了，但在服务端同样设置后并没有被blocked依旧可以正常访问。

客户端把b站的流量给拦截了吧,如果想测试,先把客户端全局代理打开,再去服务端看日志,还有blocked也要配一下.

**Reply by starllll** [reply]
我加了这段使用tiktok会被阻止

**Reply by Weathercold** [reply]
屏蔽`geoip:cn`后还需要屏蔽`geosite:cn`吗

**Reply by ArcCal** [reply]
显然不需要。也不能。

**Reply by Dr-Incognito** [reply]
关于不要使用代理访问境内 IP 网站的问题，仅针对国内而言对吧。如果人在海外，希望通过海外的VPS（例如HK、JP）加速访问国内的网站是没有关系的对吧，因为GFW没办法定位到是谁。

**Reply by Weathercold** [reply]
> 关于不要使用代理访问境内 IP 网站的问题，仅针对国内而言对吧。如果人在海外，希望通过海外的VPS（例如HK、JP）加速访问国内的网站是没有关系的对吧，因为GFW没办法定位到是谁。

没错

**Reply by Dr-Incognito** [reply]
> > 关于不要使用代理访问境内 IP 网站的问题，仅针对国内而言对吧。如果人在海外，希望通过海外的VPS（例如HK、JP）加速访问国内的网站是没有关系的对吧，因为GFW没办法定位到是谁。
> 
> 没错

对 我想了一下。代理流量根本不会经过GFW，在经过GFW的时候已经解包了。
